apollo-router/src/test_harness/http_snapshot.rs

```diff
@@ -0,0 +1,520 @@
+                                .unwrap_or(true)
+                        }),
+                        body: response_json_body,
+                    },
+                };
+                {
+                    let mut snapshots = state.snapshots.lock().unwrap();
+                    snapshots.push(snapshot.clone());
+                    if let Err(e) = save(state.snapshot_file, snapshots.as_mut()) {
+                        error!(
+                            url = %uri,
+                            method = %method,
+                            error = ?e,
+                            "Unable to save snapshot"
+                        );
+                    }
+                }
+                return Ok(snapshot.try_into().unwrap());
+            } else {
+                let message = "Unable to parse response body as JSON".to_string();
+                error_message = message.clone();
+                error!(
+                    url = %uri,
+                    method = %method,
+                    message,
+                )
+            }
+        } else {
+            let message = "Unable to read response body".to_string();
+            error_message = message.clone();
+            error!(
+                url = %uri,
+                method = %method,
+                message
+            )
+        }
+    }
+
+    Ok(http::Response::builder()
+        .status(500)
+        .body(json!({ "error": error_message}).to_string().into())
+        .unwrap())
+}
+
+fn map_headers<F: Fn(&str) -> bool>(
+    headers: HeaderMap<HeaderValue>,
+    include: F,
+) -> IndexMap<String, Vec<String>> {
+    headers.iter().fold(
+        IndexMap::new(),
+        |mut map: IndexMap<String, Vec<String>>, (name, value)| {
+            let name = name.to_string();
+            if include(&name) {
+                let value = value.to_str().unwrap_or_default().to_string();
+                map.entry(name).or_default().push(value);
+            }
+            map
+        },
+    )
+}
+
+fn save<P: AsRef<Path>>(path: P, snapshots: &mut Vec<Snapshot>) -> Result<(), SnapshotError> {
+    let path = path.as_ref();
+    if let Some(parent) = path.parent() {
+        std::fs::create_dir_all(parent)?;
+    }
+
+    // Keep the snapshots sorted to make source control merges easier when updating snapshots
+    snapshots.sort_by_key(|snapshot| {
+        format!(
+            "{}-{}",
+            &snapshot
+                .request
+                .method
+                .as_ref()
+                .unwrap_or(&String::from("GET")),
+            &snapshot.request.path.as_ref().unwrap_or(&String::from("/"))
+        )
+    });
+
+    std::fs::write(path, serde_json::to_string_pretty(&snapshots)?).map_err(Into::into)
+}
+
+fn load<P: AsRef<Path>>(path: P) -> Result<Vec<Snapshot>, SnapshotError> {
+    let str = std::fs::read_to_string(path)?;
+    let result: Result<Vec<Snapshot>, SnapshotError> =
+        serde_json::from_str(&str).map_err(Into::into);
```

Comment:
I think that you can omit the `Result::map_err` here since you mark the `SnapshotError` with [`thiserror::from`](https://docs.rs/thiserror/latest/thiserror/#details).

```suggestion
        serde_json::from_str(&str)?;
```

Reply:
Oh, I think I see that you wanted to preserve the error vs. early returning since you optionally log later. Maybe have the early return with `?` and then you don't need the conditional afterwards? Just end with `Ok(result)`